[RFC] Make inccomand work with multiline substitutions

KillTheMule opened the pull request:
Inccomand substitution now properly works with multiline matches and substitutions, handles multibyte stuff and does highlighting.

-  [x] Deal with the code duplication for highlighting without waiting for anyone, should simplify the logic anyways
-  [x] Merge the new tests into `inccommand_spec.lua`
-  [x] Lint
-  [x] Reenable timeout and its test

TODO for later:
- Work on the conflict with search highlighting (see https://github.com/neovim/neovim/pull/7315)
- Deal with code duplication (see the TODO's, wait for a reviewers opinion on that)

Closes https://github.com/neovim/neovim/issues/5608.

KillTheMule committed:
Inccommand: Multiline substitutions, highlighting, multibyte.

Make inccomand work with multiline patterns and substitutions. Also care
for proper highlighting and multibyte characters.

KillTheMule commented:
So, the approach of having a `char_u *text` in a `SubResult` does not seem appropriate. I'll make it a vec of lines, that should streamline things a bit.

KillTheMule commented:
Ok, this looks much better, though it's still wrong (I add too many matches somehow, and the preview window isn't shown). Still going :)

KillTheMule commented:
How the hell did the previous incarnation pass the tests on appveyor O_o

Anyways, logic seems sortof solid now, expect for that special casing https://github.com/neovim/neovim/pull/7231/files#diff-b9d70f1e57978ae9fbf83f73315b0c0fR3690. Next step will be to try to display the correct things in the preview window.

KillTheMule commented:
I did try some things, and now I've settled on a design I like. I'll do a bit of description below, if anyone does not think it appropriate please let me know. Status right now: Builds, works in simple manual tries, previews, but does not yet highlight. I'll first need to try some more elaborate substitutions, and then work on highlighting.

Design: I do as little as possible in `do_sub`, so performance isn't as much impacted for non-preview or even nonsplit. I only save the numbers of the lines I will need to fetch from the original buffer, as well as the position data where the match starts and ends (adjusted, because we will mostly fetch lines _after_ the substitution). Most (hopefully all) information is already there, it just needs to be saved and passed on to the preview function. In the preview function, we fetch the right lines from the original buffer (after the substitution) and print them out. The positional data for the matches will be needed for the highlighting only, but this will also make it easy to add highlighting to the original buffer for the substituted text.

Of course, comments welcome :)

KillTheMule commented:
Ok, the code is a mess, it will kindof spam your log file... BUT it's working, as far as I can see. Highlighting can be pretty weird with complex substitutions, but as far as I can see, it's correct.

So, I'd be happy for anyone to try this. I'd especially be interested in "Multi"-Situations: Multiple matches on a line, multiple-line-matches, multiple-line-substitutions, multiple-size-machtches, multiple-size-replacements, you name it!

Really though, the most useful would be a report from someone using Multibyte chars, I'll basically have to research how to input them to begin with...

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3168,7 +3177,13 @@ static buf_T *do_sub(exarg_T *eap, proftime_T timeout)
+  static linenr_T pre_started;
+  static linenr_T pre_ended;
+  static int pre_src_id; // initialized to 0, see
+                         // https://stackoverflow.com/questions/13251083
```
could be explicit, and avoid the comment :)

    static int pre_src_id = 0;

justinmk replied:
Usually we put static locals at the very start of a function. I think it would be surprising to see them in the middle of other locals.

KillTheMule replied:
About that explizit, I'm not sure about the semantics. I don't want it to be reset to 0 every time the function is called, of course. That won't happen?

justinmk replied:
@KillTheMule won't happen. static locals are initialized only once for the life of the process.

KillTheMule replied:
Which, thinking about it, makes sense! Thanks, I'll change that.

KillTheMule committed:
Fix old inccomand tests, and add more for the new functionality.

KillTheMule commented:
Fixed the old tests and an outstanding issue. Also started writing new tests.

I'm baffled that the new testshttps://github.com/neovim/neovim/pull/7231/files#diff-ddcc7112118dfbcc54565044d90602eb also pass on master. When I try this manually on current master, nvim locks up and the substitution does not happen, so the expect should fail. But the tests do pass...

KillTheMule commented:
Fixed a lot of bugs during writing the tests. Basically it is polished enough for a review, although I definitely want to simplify the highlighting logic in `show_sub` soon.

If anyone can think of another test to be added, I'd be happy to hear it.

KillTheMule committed:
Lint

KillTheMule committed:
Move tests into original file, lint, and add a test

KillTheMule commented:
Ok, that's it from my POV, there's one code duplication left, but justinmk told me to leave it for now, it's not the end of the world indeed.

I'm bad at naming things (variables, functions, structures) and making comments, bikeshedding definitely welcome.

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -6028,6 +6089,8 @@ static buf_T *show_sub(exarg_T *eap, pos_T old_cusr, char_u *pat, char_u *sub,
   char_u *save_shm_p = vim_strsave(p_shm);
   size_t sub_size = mb_string2cells(sub);
   size_t pat_size = mb_string2cells(pat);
```
This was another case where I apparently hyper-corrected from strlen() to mb_string2cells() in c04ffe866d276d6a6bd9e9c6a8b0dbb71504db7c . I think it may not be correct here, also. At least when `pat_size` is an argument to `bufhl_add_hl()` (below), that seems wrong. But it's also used to compute column widths in the "highlight the replaced part" section below ...

KillTheMule replied:
Those are not used for anything anymore other than the decision if we need to show the split window, see https://github.com/neovim/neovim/pull/7231/files#diff-b9d70f1e57978ae9fbf83f73315b0c0fR6105.  We only check if those are 0 or not, so `mb_string2cells` or `strlen` should not matter (actually, we could really check `*sub` and `*pat`).

I'm not sure why that check is there anyways. If `pat` or `sub` are empty or `NULL`, doesn't that just mean the user left it out and the last recent pattern/substitution is used? Why wouldn't we want to show the preview window then?

I think we could just remove those 2 lines, remove this part of the check and remove those arguments from the function call.

justinmk replied:
> I'm not sure why that check is there anyways. If pat or sub are empty or NULL, doesn't that just mean the user left it out and the last recent pattern/substitution is used?

Well, it's a matter of whether we think the user wants to see a preview of `:%s/`. At the time I guess this would be more annoying than helpful.

> I think we could just remove those 2 lines, remove this part of the check and remove those arguments from the function call.

👍  Let's put it in a separate commit for clarity.

KillTheMule replied:
Hmm, right, maybe the user doesn't want that preview... I'll ponder it a bit and prepare a new commit tomorrow, this should be streamlined no matter what.

justinmk replied:
I think the simplification is worth it.

KillTheMule committed:
Remove pat/sub from show_sub

They were only used to not show the preview window when typing "s/" or
"s//" only, in which case the previous pattern would be reused. Now the
window is shown in that case.

KillTheMule commented:
I've pushed the mentioned simplification, I like it better this way, although it shows our wonky highlighting before the second delimiter once more. Need to work on that soon :)

I've also cooked up another simplification. Let me know if I should leave that as a different commit or if I should squash that into the first one.

KillTheMule committed:
Fix the last line in the preview buffer

It would always show an empty line at the end that didn't belong.

KillTheMule committed:
Adjust tests for the new preview window

... that does not have that superflous last line.

Also, remove some indeterminism for the freebsd64 tests. Partially,
those were suggested by the tests themselves, while successfull. Some of
them were added after some testing because the lookaround test would
fail on freebsd64 only.

KillTheMule commented:
I've added another little fix: The preview buffer would always show one superflous line (I guess that was the one that's added to a buffer upon creation). I removed that, and therefore had to adjust quite some tests.

KillTheMule commented:
Note to self: Need to clear highlight if there's no more match.

KillTheMule committed:
Clear highlight when there's no match

KillTheMule committed:
Test for clearing the highlight

justinmk commented:
Merged in https://github.com/neovim/neovim/pull/7315
